### ISSUE TITLE
Fetch adds charset=utf-8 on Android but not on iOS

### ISSUE THREAD

**Issue by dstaver** [facebook/react-native#8237]
This fetch request sends "application/json" as ContentType on IOS but "application/json; charset=utf-8" on Android:

```
fetch(url, {
    method: 'POST',
    headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
    }
})
```

The server I'm working against requires application/json and fails when charset is added. I've filed a bug report with them too, but shouldn't fetch behave identically on ios and android?

**Comment by dstaver** [comment]
http://stackoverflow.com/questions/25560601/how-to-suppress-charset-being-automatically-added-to-content-type-in-okhttp

**Comment by israelidanny** [comment]
I'm also experiencing this. Having different headers between platforms doesn't make sense and introduces bugs

**Comment by lacker** [comment]
What's the right thing to do here - should the charset be appended or not?

**Comment by himelnagrana** [comment]
Having the same issue while working with Urban Airship APIs. Frustrating as UA supposed to be not changing at their end. What to do?

**Comment by Negan1911** [comment]
Im also seeing this

**Comment by ide** [comment]
> What's the right thing to do here - should the charset be appended or not?

Looking at Chrome, it doesn't modify the content-type header at all so we shouldn't be adding a charset.

**Comment by alterx** [comment]
I'm also experiencing this issue, somehow `charset=utf-8` is added as a parameter to the `Content-Type` header. It only happens on Android.

In my specific case, I'm interacting with a JSON API (http://jsonapi.org/format/) and adding media type parameters causes the server to respond with a 415 error every time. This pretty much breaks the app :(

**Comment by laurent22** [comment]
I'm also having this problem and, while doing some testing, I've tried setting an invalid charset "dropbox-cors-hack" (as I'm testing the Dropbox API) so the request was like this:

    apiRequest = request.post(getBaseURL(host) + path)
      //.type('application/octet-stream;')
      .type('text/plain; charset=dropbox-cors-hack')
      .set('Authorization', 'Bearer ' + accessToken)
      .set('Dropbox-API-Arg', httpHeaderSafeJson(args));

When doing so I got the following exception:

![](https://i.imgur.com/Ii4J0t7.png)

And that lead me to this file from the okhttp library:

https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/RequestBody.java#L53

So it seems it appends "charset=utf-8" if the charset is not set. I'm not sure if this is React-Native that is calling this lib or if it's Android, but is there anything that can be done to prevent this behaviour?

**Comment by laurent22** [comment]
Based on this comment it doesn't look like it will be fixed on the okhttp side:

https://github.com/square/okhttp/issues/3081

They suggest using the `create` overload that takes a `byte[]` parameter. Can it be done at the React-Native level?

**Comment by Knight704** [comment]
@laurent22 you could customize okhttp client on android that is used under the hood to implement fetch. You can add interceptor to this client and modify any header before it being processed.
Unfortunately, current version on react-native doesn't provide an easy way to provide your own okhttp client, but [here](https://github.com/facebook/react-native/issues/13487#issuecomment-319984503) I posted a way to workaround this.

**Comment by Knight704** [comment]
@hey99xx the idea boils down to put your custom interceptor to okhttpclient. Yes, okhttp will continue appending charset into header, but inside interceptor you can modify any headers before it really goes to the network, i.e cutting down charset part. This is of course a workaround, but it'll work

**Comment by hey99xx** [comment]
@Knight704 How would you know if the "Content-type" is automatically set to have the charset part by OkHttp, and not by manually adding the header during `fetch` call? I mean it's an acceptable solution, but it comes with few questions. Ideally it should be fixed in one of react-native or OkHttp so we happily keep developing our apps. 

On my end, I've managed to fix the issue by contacting the backend people to be tolerant to the charset header, but I don't think that's doable when you contact external APIs such as AWS and Dropbox.

**Comment by Knight704** [comment]
@hey99xx yeah, I've already mentioned that this is just a hack, and preferably it should be fixed in SDK.
Nevertheless, if you are developing your app you should know endpoints where charset should (or shoudn't) be added and may have black/white list that is checked inside interceptor, there are many ways for this, the first that comes to my mind is putting your custom header like 'no-charset-append' on js side and cut down this appending if this header exists on native side. Ugly, but flexible and would work

**Comment by jeremy8883** [comment]
Following from the steps outlined [here](https://github.com/facebook/react-native/issues/13487#issuecomment-319984503) (thanks heaps for that by the way @Knight704), for anyone wanting a copy/paste example, my version of `NetworkingModuleUtils` looked like this:

```
package com.facebook.react.modules.network;

import com.facebook.react.bridge.ReactApplicationContext;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class NetworkingModuleUtils {
    public static NetworkingModule createNetworkingModuleWithCustomClient(ReactApplicationContext context) {
        OkHttpClient client = OkHttpClientProvider.createClient();

        OkHttpClient customClient = client.newBuilder()
            .addNetworkInterceptor(new Interceptor() {
                @Override
                public Response intercept(Chain chain) throws IOException {
                    Request request = chain.request();
                    String contentType = request.header("Content-Type");
                    Request customRequest = "application/json; charset=utf-8".equals(contentType) ?
                        request.newBuilder()
                            .header("content-type", "application/json")
                            .build() : request;

                    Response response = chain.proceed(customRequest);
                    return response;
                }
            }).build();

        return new NetworkingModule(context, null, customClient);
    }
}
```

**Comment by Kottidev** [comment]
@jeremy8883,   in which directory i copy and past your code ? in src/java/com/ ?

your class resolve the problem for appending "charset=utf-8" to content-type ?

**Comment by jeremy8883** [comment]
Yeah, plus the package location, so `src/main/java/com/facebook/react/modules/network/NetworkModuleUtils.java`. Make sure you also follow the steps in [this link](https://github.com/facebook/react-native/issues/13487#issuecomment-319984503) also.

**Comment by ceefour** [comment]
I use `Content-Type: image/jpeg` and it becomes `Content-Type: image/jpeg; charset=utf-8` which obviously doesn't make sense. :(

**Comment by laurent22** [comment]
Please don't close, this is still an issue in latest RN

**Comment by nateReiners** [comment]
+1

**Comment by israelidanny** [comment]
This still breaks some requests, please don't close it!

**Comment by alien3d** [comment]
react-native version :0.53
```
03-03 10:14:35.084  8880  9189 E UncaughtException: 
03-03 10:14:35.084  8880  9189 E UncaughtException: java.nio.charset.UnsupportedCharsetException: utf8mb4
```

**Comment by israelidanny** [comment]
This is still an issue in lates RN , please don’t close

**Comment by archer56** [comment]
+1

**Comment by limbuster** [comment]
This is still an issue

**Comment by jeremy8883** [comment]
Axios is a javascript library. The problem is in the RN/okhttp java implementation.

**Comment by programadornatal** [comment]
Axios work for me in IOS

**Comment by ceefour** [comment]
@programadornatal iOS doesn't have this bug even using pure fetch, without using axios. This is Android specific.

**Comment by glenndebacker** [comment]
Axios has the same problem on Android. There is a topic describing the same problem but the solution of changing the default header does not seem to work. Android (or RN) seems to be modifying it on a lower level.

**Comment by nhunzaker** [comment]
This seemed like a fun one to dig into, so I proposed a fix here:
https://github.com/facebook/react-native/pull/23580

**Comment by hey99xx** [comment]
@nhunzaker I know I approved your PR but just looking again, I wanted to clarify this. I cannot ask on the PR itself since it's locked already.

You had mentioned

> I wondered about that, or passing in the character set as defined by `MediaType.parse`.

Do you still think it'd be better? One use case that came to my mind is setting a custom content-type header like: `Content-Type: text/plain; charset=utf-16`. 

In that case OkHttp's `RequestBody.create` would have likely used utf-16 charset to get the bytes. Now after this change we'd be sending bytes encoded with utf-8, but the charset in the header does not match, potentially confusing the servers.

I really don't know what the correct implementation is supposed to be. Maybe only default to utf-8 is MediaType is null or charset-less? [That's what OkHttp internally does.](https://github.com/square/okhttp/blob/parent-3.10.0/okhttp/src/main/java/okhttp3/RequestBody.java#L43-L58)

**Comment by hey99xx** [comment]
Example of how this failure could be> I've never written a raw http server though so cant tell if this is how it works:

```
String s = "Friðjónsson";
Charset UTF_8 = StandardCharsets.UTF_8;
Charset UTF_16 = StandardCharsets.UTF_16;
System.out.println("original: " + s);
System.out.println("utf8: " + new String(s.getBytes(UTF_8), UTF_8));
System.out.println("utf16: " + new String(s.getBytes(UTF_16), UTF_16));
System.out.println("utf8/16: " + new String(s.getBytes(UTF_8), UTF_16));
```

output is:

```
original: Friðjónsson
utf8: Friðjónsson
utf16: Friðjónsson
utf8/16: 䙲槃끪쎳湳獯�
```

**Comment by nhunzaker** [comment]
@hey99xx you are correct. my implementation needs to respect the existing character set. I have a test case that reproduces the issue, I'll send out another PR. The code I changed needs to really be something like:

```
Charset charset = contentMediaType.charset();

if (charset == null) {
  charset = StandardCharsets.UTF_8;
}

requestBody = RequestBody.create(contentMediaType, body.getBytes(charset));
```

**Comment by hey99xx** [comment]
The fixes are in 0.59.0-rc.3 which is just released. Can someone verify the problem is gone and we can then resolve this issue?

**Comment by bartolkaruza** [comment]
Confirmed that this works on RN 0.59.1!
